docs/libFuzzer-integration.md

```diff
@@ -0,0 +1,184 @@
+    ReadBuffer(discard_buffer, num_to_discard);
+    num_discarded += num_to_discard;
+  }
+}
+
+struct __attribute__((packed)) MutateRequestHeader {
+  uint64_t size;
+  uint32_t seed;
+  uint8_t is_fragment;
+};
+
+static struct sockaddr_in serv_addr;
+extern "C" size_t LLVMFuzzerCustomMutator(uint8_t* data, size_t size,
+                                          size_t max_size, unsigned int seed) {
+  if (size <= 1) {
+    // Handle common invalid testcases gracefully.
+    static const std::string basic_shader = "void main(void) { }";
+    if (basic_shader.size() < max_size) {
+      memcpy(reinterpret_cast<char*>(data), basic_shader.c_str(),
+             basic_shader.size());
+      return basic_shader.size();
+    }
+  }
+
+  // Open a connection to the CustomMutatorServer.
+  if (!sock) {
+    sock = socket(AF_INET, SOCK_STREAM, 0);
+    CHECK(sock >= 0, "Could not create socket");
+    static struct sockaddr_in serv_addr;
+    serv_addr.sin_family = AF_INET;
+    static const int port = 8666;
+    serv_addr.sin_port = htons(port);
+    CHECK(inet_pton(AF_INET, "0.0.0.0", &serv_addr.sin_addr) != -1,
+          "invalid address");
+
+    CHECK(connect(sock, reinterpret_cast<const struct sockaddr*>(&serv_addr),
+                  sizeof(serv_addr)) >= 0,
+          "connection failed");
+  }
+
+  static MutateRequestHeader request_header;
+  request_header.size = size;
+  request_header.seed = seed;
+  // In this example we only start with a fragment shader, so every shader must
+  // be a fragment shader.
+  request_header.is_fragment = true;
+  static int flags = 0;
+  // Send the mutation request.
+  ssize_t num_bytes_sent =
+      send(sock, reinterpret_cast<const void*>(&request_header),
+           sizeof(request_header), 0);
+  CHECK(num_bytes_sent == sizeof(request_header), "short write");
+  // Send the shader.
+  num_bytes_sent = send(sock, reinterpret_cast<const void*>(data), size, flags);
+  CHECK(static_cast<size_t>(num_bytes_sent) == size, "short write");
+
+  // Read the response shader size and contents.
+  uint8_t mutated_size_bytes[sizeof(size_t)];
+  ReadBuffer(mutated_size_bytes, sizeof(size_t));
+  size_t mutated_buf_length = *(reinterpret_cast<size_t*>(mutated_size_bytes));
+  if (mutated_buf_length > max_size) {
+    Discard(mutated_buf_length);
+    std::cout << "discard" << std::endl;
+    return size;
+  }
+  size_t read_size = std::min(max_size, mutated_buf_length);
+  ReadBuffer(data, read_size);
+  return read_size;
+}
+
+extern "C" int LLVMFuzzerTestOneInput(const uint8_t* data, size_t size) {
+  std::string shader(reinterpret_cast<const char*>(data), size);
+  // Give the user some feedback since the coverage won't grow much in an empty
+  // fuzzer.
+  std::cout << shader << std::endl;
+  return 0;
+}
+EOF
+
+# Must use clang-6.0 or greater (sudo apt install clang-6.0 && clang-6.0 -fsanitize=fuzzer...).
+clang++ -g -fsanitize=fuzzer,address fuzzer.cc -o fuzzer
```

Comment:
By the way, this command fails on my machine, I need to use `clang-6.0` instead of clang++, but I think the clang on our boxes is older than most.

Reply:
Note for us (I doubt we need to add this to the docs): As well as installing `clang-6.0`, install `libstdc++-8-dev`.